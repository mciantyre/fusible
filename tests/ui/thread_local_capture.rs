// SPDX-License-Identifier: MPL-2.0
// SPDX-FileCopyrightText: Copyright 2024 Ian McIntyre

use core::pin::pin;
use fusible::thread::{StaticStack, Thread};

static STACK: StaticStack<512> = StaticStack::new();

/// By default, a thread's entrypoint cannot capture local state.
fn main() {
    let x = 5;
    let thread = pin!(Thread::context());
    Thread::create(
        thread.into_ref(),
        STACK.take().unwrap(),
        &Default::default(),
        || {
            let _nope = &x;
            let _fine = &STACK;
        },
    )
    .unwrap();
}
