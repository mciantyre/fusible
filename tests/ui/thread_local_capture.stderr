error[E0373]: closure may outlive the current function, but it borrows `x`, which is owned by the current function
  --> tests/ui/thread_local_capture.rs:17:9
   |
17 |         || {
   |         ^^ may outlive borrowed value `x`
18 |             let _nope = &x;
   |                          - `x` is borrowed here
   |
note: function requires argument type to outlive `'static`
  --> tests/ui/thread_local_capture.rs:13:5
   |
13 | /     Thread::create(
14 | |         thread.into_ref(),
15 | |         STACK.take().unwrap(),
16 | |         &Default::default(),
...  |
20 | |         },
21 | |     )
   | |_____^
help: to force the closure to take ownership of `x` (and any other referenced variables), use the `move` keyword
   |
17 |         move || {
   |         ++++
